parser/lexer: bump to Unicode 17, use faster unicode-ident

Marcondiro opened the pull request:
Hello,

Bump the unicode version used by lexer/parser to 17.0.0 by updating:
- `unicode-normalization` to 0.1.25
- `unicode-properties` to 0.1.4
- `unicode-width` to 0.2.2

and by replacing `unicode-xid` with `unicode-ident` which is also 6 times faster.
I think it might be worth to run the benchmarks to double check.
(`unicode-ident` is already in `src/tools/tidy/src/deps.rs`)

Thanks!

Marcondiro committed:
parser/lexer: bump to Unicode 17, use faster unicode-ident

Replace unicode-xid with unicode-ident which is 6 times faster

rust-timer commented:
Finished benchmarking commit (988451ce73b832a095adca69acf309ce27a2f54d): [comparison URL](https://perf.rust-lang.org/compare.html?start=23c7bad921fb7163de37ea680bed317deaa03fda&end=988451ce73b832a095adca69acf309ce27a2f54d&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=23c7bad921fb7163de37ea680bed317deaa03fda&end=988451ce73b832a095adca69acf309ce27a2f54d&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.1%, 0.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=23c7bad921fb7163de37ea680bed317deaa03fda&end=988451ce73b832a095adca69acf309ce27a2f54d&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=23c7bad921fb7163de37ea680bed317deaa03fda&end=988451ce73b832a095adca69acf309ce27a2f54d&stat=cycles:u)
<details>
<summary>Results (primary 2.6%, secondary 3.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.6% | [2.6%, 2.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | 3.6% | [3.6%, 3.6%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.6% | [2.6%, 2.6%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=23c7bad921fb7163de37ea680bed317deaa03fda&end=988451ce73b832a095adca69acf309ce27a2f54d&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 473.971s -> 474.835s (0.18%)
**Artifact size**: 390.89 MiB -> 390.89 MiB (-0.00%)

joshtriplett commented:
I think it needs *some* level of review, to make sure that (for instance) the new Unicode version isn't doing anything out of the ordinary, and to make sure that *some* person in the project experienced with Unicode has taken at least a cursory look at the changes to `XID_Start`/`XID_Continue` and any related changes to confusables that overlap with `XID_Start`/`XID_Continue`.

I *don't* think that review is best done in lang. I think we should delgate that to whichever team is making sure of the above. (Is that T-compiler?) So, ideally, I'd love to see a proposal to lang requesting a delegation to take responsibility for the above.

That said, let's go ahead and sign off on *this* change to unblock it.

bors commented:
:umbrella: The latest upstream changes (presumably #147029) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
r? @joshtriplett 

I tend to agree that FCP seems like the wrong tool to confirm implementation details (e.g., membership of XID sets). I don't know what specific review you'd like to perform so yielding to you.

Mark-Simulacrum commented:
Personally, I'm happy to r+ this, but several people on lang indicated they wanted some kind of review. It's not obvious to me whether their check boxes on the FCP indicate they have done said review (it's not clear to me what would go into that, so I can't really check).

It's not clear to me what would block us from bumping the Unicode version we're using (maybe I don't know enough, though), so in light of that uncertainty I didn't want to just r+. But if we treat the FCP as having resolved it then I'm happy to effectively rubber stamp this.

@ehuss, you had some concerns over using different crates here - happy for us to not approve this for those reasons (I don't fully understand that concern but haven't dug into the details).

Mark-Simulacrum commented:
> Since all the relevant crates seem to export some flavour of the UNICODE_VERSION constant included in the standard library, perhaps there could be a simple test somewhere that asserts that all these constants are identical? That way manual review of the versions isn't required.

This sounds like a good idea. Can we add a static assert or similar on this?

> No concerns have been filed, and our FCP is complete [...]

Given the above I think I'm OK moving forward with the asserts added. r=me with that done.

Marcondiro commented on tests/ui-fulldeps/lexer/unicode-version.rs:
> Since all the relevant crates seem to export some flavour of the UNICODE_VERSION constant included in the standard library, perhaps there could be a simple test somewhere that asserts that all these constants are identical? That way manual review of the versions isn't required.

At the moment there is this test in place, checking that `rustc_lexer::UNICODE_IDENT_VERSION` == `rustc_parse::UNICODE_NORMALIZATION_VERSION`

Would expanding this test to all the Unicode-version dependent crates used in `/compiler/*` be enough?

clarfonthey commented:
Gonna preface this with: my main reason for having this discussion in the first place is, as I mostly mentioned, I think that bumping the Unicode version, since it is a routine change, should have some sort of playbook so we don't need to have these discussions again. And, so, I'm having these discussions here because I think the formation of said playbook is relevant, but if you'd rather move it somewhere else, I'm fine doing that too.

> Note that this is mostly because unicode_rs crates get bumped when someone wants them to be. If Rust wanted them to be bumped, they can be bumped (best way: make a PR with the bump).

Not that this is the place to put this, but the exact reason why I submitted unicode-rs/unicode-script#25 (which you merged) is because Clippy uses `unicode-script` and it doesn't have a published version for Unicode 17, so, an attempt to enforce that all the Unicode versions match for clippy specifically fail. I imagine, then, if we were to enforce all of this, the procedure would then be to ping all of the authors of these crates, ensure that they're all updated to the latest version of Unicode, then perform the bump for all R-L crates? While this is also a request to do so for this particular crate, I'm also using it as an example since this would be part of what people should do going forward.

Right now, tidy only allows checking particular crates in `Cargo.lock`, and not explicit versions of them, so, we can't particularly enforce only versions which support the right Unicode version. This is a potential side concern of verifying the Unicode version is consistent: even though we can check the `UNICODE_VERSION` const for everything we use, we can't verify that dependencies also use the same version, which, even if it doesn't matter *that* much, is still probably better for the sake of reducing table size in binaries. So, I guess, that's also a side note here. I'm not sure exactly what the policy is for allowing dependencies in the compiler, but I assume that at least part of it is some assurance that the people who maintain them are at least active enough to merge changes if the compiler needs it.

There's also a potential consideration of just including the Unicode functionality most critical to the compiler directly in the compiler itself. Since all of the code used has a compatible license, it shouldn't be tremendously difficult to copy it in somewhere, and it would also mean that we could share some of the code that parses stuff like `UnicodeData.txt` between these various crates. This isn't to say it should be part of libstd, but maybe having it in the compiler to make version consistency easier is a consideration.

Marcondiro committed:
lexer/parser: ensure deps use the same unicode version

Add a compile time check in rustc_lexer and rustc_parse ensuring that unicode-related dependencies within the crate use the same unicode version.
These checks are inspired by the examples privided by @clarfonthey.

crlf0710 commented:
> I think that bumping the Unicode version, since it is a routine change, should have some sort of playbook so we don't need to have these discussions again

#101840 could be an outdated (year 2022) example of such a playbook.

Mark-Simulacrum commented:
@bors r+ rollup
